ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java

```diff
@@ -147,39 +157,60 @@ void fetchAndRunCommands() {
 
     final Optional<QueuedCommand> terminateCmd = findTerminateCommand(commands);
     if (terminateCmd.isPresent()) {
-      terminateCluster(terminateCmd.get().getCommand());
+      terminateCluster(terminateCmd.get());
       return;
     }
 
     log.trace("Found {} new writes to command topic", commands.size());
-    for (final QueuedCommand command : commands) {
-      if (closed) {
-        return;
+    final Iterator<QueuedCommand> commandIt = commands.iterator();
+    synchronized (executorLock) {
+      while (commandIt.hasNext() && !closed) {
+        highestExecutedOffset = executeStatement(commandIt.next());
       }
-
-      executeStatement(command);
     }
   }
 
-  private void executeStatement(final QueuedCommand queuedCommand) {
-    log.info("Executing statement: " + queuedCommand.getCommand().getStatement());
+  private long executeRestore(final QueuedCommand queuedCommand) {
+    return executeStatement(queuedCommand, () -> statementExecutor.handleRestore(queuedCommand));
+  }
 
-    final Runnable task = () -> {
+  private long executeStatement(final QueuedCommand queuedCommand) {
+    return executeStatement(queuedCommand, () -> {
       if (closed) {
         log.info("Execution aborted as system is closing down");
       } else {
         statementExecutor.handleStatement(queuedCommand);
         log.info("Executed statement: " + queuedCommand.getCommand().getStatement());
       }
-    };
+    });
+  }
 
-    RetryUtil.retryWithBackoff(
-        maxRetries,
-        STATEMENT_RETRY_MS,
-        MAX_STATEMENT_RETRY_MS,
-        task,
-        WakeupException.class
-    );
+  @GuardedBy("executorLock")
+  private long executeStatement(final QueuedCommand queuedCommand, final Runnable task) {
+    if (!hasValidChecksum(queuedCommand)) {
+      log.warn("Not executing command {} because the checksum is invalid!", queuedCommand);
+    } else {
+      log.info("Executing statement: " + queuedCommand.getCommand().getStatement());
+      RetryUtil.retryWithBackoff(
```

Comment:
I agree, but I don't know if that's feasible today. We throw `KsqlException` for a bunch of random things (e.g. KafkaTopicClient fails to do something).

I think it would be a good exercise to go through our codebase and get a more rigorous definition of exceptions, and make sure we are handling them properly, but that will take quite a bit of time.

Reply:
I mean we shouldn't ever ignore any exceptions when processing statements. There should be no expected exceptions, so we shouldn't move past a statement unless its checksum is wrong or we succeed to execute it.